### PR TITLE
feat: add degrees data loader

### PR DESCRIPTION
**Pull request by azanbinzahid**
## Description:
This PR adds Degrees CSV loader to populate incoming degrees' data in `course_metadata` app.

## Linked Ticket:
[PROD-2807](https://2u-internal.atlassian.net/browse/PROD-2807)

## Testing Instructions
- Checkout masters, run migrations.
- Checkout this branch
- Run tests `pytest course_discovery/apps/course_metadata/data_loaders/tests/test_degree_csv_loader.py --ds=course_discovery.settings.test_local --reuse-db`
- Run loader ` ./manage.py import_degree_data --csv_path=./degrees.csv`
- Test file: [degrees.csv](https://github.com/openedx/course-discovery/files/8814774/degrees.csv)
- Setup Subject, LevelType (and its translation) in django admin

## TODO:
- [x] refactor
- [x] update and rebase after #3428 
- [x] update and rebase after #3444 
- [x] update and rebase after #3441
- [x] fix existing degree update test failure
- [x] Write pending method 
- [x] Move loader to separate file
- [x] Write unit tests
- [x] complete TODO's in PR
- [ ] decide unique key after title discussion
- [ ] Write validator

### CODE REVIEW

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+                    degree_title, message
+                ))
+                continue
+
+            # TODO: Handle Card Image
+
+            # get degree object from title
+            degree = Degree.objects.filter(title=degree_title, partner=self.partner).first()
+
+            degree_title = degree.title if degree else degree_title
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/data_loaders/degrees_loader.py]: why is this needed?

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+        Make a degree object through ORM
+        """
+        org = Organization.objects.get(key=data['organization_key'])
+
+        degree = Degree.objects.create(
+            title=data['title'],
+            type=program_type,
+            status=ProgramStatus.Unpublished,
+            marketing_slug=data['slug'],
+            overview=data['overview'],
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/data_loaders/degrees_loader.py]: Is overview populating when creating the degree?

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+        """
+        transformed_dict = {}
+        for key, value in data.items():
+            updated_key = key.strip().lower().replace(' ', '_')
+            transformed_dict[updated_key] = value
+        return transformed_dict
+
+    def _create_degree(self, data, program_type):
+        """
+        Make a degree object through ORM
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/data_loaders/degrees_loader.py]: once all the pieces are in, update the docstring to include what operations/data models associations are happening during creation

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+            logger.info(updated)
+        else:
+            logger.error("Degree with title {} has not been updated".format(    # lint-amnesty, pylint: disable=logging-format-interpolation
+                degree,
+            ))
+            return
+
+        if degree.authoring_organizations.filter(key=data['organization_key']).count() == 0:
+            # degree.authoring_organizations.clear() TODO
+            degree.authoring_organizations.add(org)
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/data_loaders/degrees_loader.py]: question: Do the attempts to add an existing org result in an error?

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py (hunk 1 above)
Comment (azanbinzahid) [reply course_discovery/apps/course_metadata/data_loaders/degrees_loader.py]: We need title for logging purpose, it has been initiated from csv row title.

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py (hunk 2 above)
Comment (azanbinzahid) [reply course_discovery/apps/course_metadata/data_loaders/degrees_loader.py]: Yes, I think I found the update issue.

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py (hunk 4 above)
Comment (azanbinzahid) [reply course_discovery/apps/course_metadata/data_loaders/degrees_loader.py]: No, it doesn't.

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+            self.reader = csv.DictReader(open(csv_path, 'r'))
+        except FileNotFoundError:
+            logger.exception("Error opening csv file at path {}".format(csv_path))    # lint-amnesty, pylint: disable=logging-format-interpolation
+            raise  # re-raising exception to avoid moving the code flow
+
+    def ingest(self):  # pylint: disable=too-many-statements
+        logger.info("Initiating Degree CSV data loader flow.")
+        for row in self.reader:
+            # TODO: is this the right way?
+            with transaction.atomic():
```
Comment (azanbinzahid) [inline course_discovery/apps/course_metadata/data_loaders/degrees_loader.py:36]: @DawoudSheraz @uzairr

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+                # TODO: Handle organizations override
+
+                program = Program.objects.get(degree=degree, partner=self.partner)
+                is_downloaded = download_and_save_program_image(program, row['card_image_url'])
+                if not is_downloaded:
+                    logger.error("Unexpected error happened while downloading image for degree {}".format(  # lint-amnesty, pylint: disable=logging-format-interpolation
+                        degree_title
+                    ))
+                    self.messages_list.append('[IMAGE DOWNLOAD FAILURE] degree {}'.format(degree_title))
+                    continue
```
Comment (azanbinzahid) [inline course_discovery/apps/course_metadata/data_loaders/degrees_loader.py:122]: should we roll back if image download fail?

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py (hunk 5 above)
Comment (azanbinzahid) [inline course_discovery/apps/course_metadata/data_loaders/degrees_loader.py]: @DawoudSheraz @uzairr is this the right way? will it rollback automatically?

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+                message = self.validate_degree_data(program_type, row)
+                if message:
+                    logger.error("Data validation issue for degree {}, skipping ingestion".format(degree_title))    # lint-amnesty, pylint: disable=logging-format-interpolation
+                    self.messages_list.append("[DATA VALIDATION ERROR] Degree {}. Missing data: {}".format(
+                        degree_title, message
+                    ))
+                    continue
+
+                # TODO: handle org override
+                # simple field populate
```
Comment (azanbinzahid) [inline course_discovery/apps/course_metadata/data_loaders/degrees_loader.py]: after #3428

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+
+                degree_title = degree.title if degree else degree_title
+
+                if degree:
+                    logger.info("Degree {} is located in the database.".format(degree_title))    # lint-amnesty, pylint: disable=logging-format-interpolation
+
+                    additional_metadata = DegreeAdditionalMetadata.objects.filter(degree=degree).first()
+
+                    if not additional_metadata or \
+                            additional_metadata.external_identifier == external_identifier:
```
Comment (azanbinzahid) [inline course_discovery/apps/course_metadata/data_loaders/degrees_loader.py]: Assumption: all new degree's will have additional_metadata from the scratch

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py (hunk 6 above)
Comment (azanbinzahid) [inline course_discovery/apps/course_metadata/data_loaders/degrees_loader.py]: Should we rollback degree addition if image download fail?

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+            logger.info("Degree UUIDs:")
+            for degree_uuid, title in self.degree_uuids.items():
+                logger.info("{}:{}".format(degree_uuid, title))    # lint-amnesty, pylint: disable=logging-format-interpolation
+
+    def validate_degree_data(self, program_type, data):  # pylint: disable=unused-argument
+        """
+        Verify the required data key-values for a program type are present in the provided
+        data dictionary and return a comma separated string of missing data fields.
+        """
+        # TODO
```
Comment (azanbinzahid) [inline course_discovery/apps/course_metadata/data_loaders/degrees_loader.py:135]: possibly followup PR

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+                degree,
+            ))
+        else:
+            logger.error("Degree with title {} has not been updated".format(    # lint-amnesty, pylint: disable=logging-format-interpolation
+                degree,
+            ))
+            return
+
+        org = Organization.objects.get(key=data['organization_key'])
+        # degree.authoring_organizations.clear() TODO
```
Comment (azanbinzahid) [inline course_discovery/apps/course_metadata/data_loaders/degrees_loader.py]: should we clear existing orgs (if any)?

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py (hunk 5 above)
Comment (DawoudSheraz) [reply course_discovery/apps/course_metadata/data_loaders/degrees_loader.py]: Yes, but with a catch. Since we are handling the exceptions, django might not know an error has occurred. 
See https://docs.djangoproject.com/en/4.0/topics/db/transactions/#django.db.transaction.atomic and the note `Avoid catching exceptions inside atomic!`. What we can do is move the atomic block below when the degrees are created/updated. If the exception move-around is a significant effort, we can do that in a follow-up PR too.

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py (hunk 10 above)
Comment (DawoudSheraz) [reply course_discovery/apps/course_metadata/data_loaders/degrees_loader.py]: It is unlikely for the orgs to change. But ok to play on the safe side.

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py (hunk 6 above)
Comment (DawoudSheraz) [reply course_discovery/apps/course_metadata/data_loaders/degrees_loader.py]: It is ok. Image download can fail for many reasons, so rolling back degree changes might not be feasible. We won't be getting into any big inconsistent data state in the case of download failure.

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py (hunk 5 above)
Comment (azanbinzahid) [reply course_discovery/apps/course_metadata/data_loaders/degrees_loader.py]: That's good insight, when I tested it wasn't rolling back on some failures because of the reason mentioned above. Will see how we can go about it.

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py (hunk 6 above)
Comment (azanbinzahid) [reply course_discovery/apps/course_metadata/data_loaders/degrees_loader.py]: I guess we can remove this `continue` then.

**Comment by DawoudSheraz** [comment]
Going to test it out on the local with WIP data.

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+            self.reader = csv.DictReader(open(csv_path, 'r'))  # lint-amnesty, pylint: disable=consider-using-with
+        except FileNotFoundError:
+            logger.exception("Error opening csv file at path {}".format(csv_path))    # lint-amnesty, pylint: disable=logging-format-interpolation
+            raise  # re-raising exception to avoid moving the code flow
+
+    def ingest(self):
+        logger.info("Initiating Degree CSV data loader flow.")
+        for row in self.reader:
+            # TODO: test and decide if need to make transaction atomic
+            # with transaction.atomic():
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/data_loaders/degrees_loader.py]: With the recent refactoring, do you think if that is needed anymore?

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py [hunk 12]
```diff
@@ -0,0 +1,283 @@
...
+            # level_type_override = self._get_object(
+            #     LevelType, "name_t",
+            #     row['level_type_override'], degree_title
+            # )
+            # language_override = self._get_object(
+            #     LanguageTag, "code",
+            #     row['language_override'], degree_title
+            # )
+
+            if not org or not program_type:
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/data_loaders/degrees_loader.py]: nit: not (org and program_type)

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py [hunk 13]
```diff
@@ -0,0 +1,283 @@
...
+                self.messages_list.append("[DATA VALIDATION ERROR] Degree {}. Missing data: {}".format(
+                    degree_title, message
+                ))
+                continue
+
+            # get degree object from title and external_identifier
+            degree = Degree.objects.filter(
+                title=degree_title, partner=self.partner,
+                # additional_metadata__external_identifier=row['identifier']
+            ).first()
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/data_loaders/degrees_loader.py:71]: I believe we can introduce this check in a follow-up PR when we have removed title uniqueness. Otherwise the same-titled degrees with different identifiers won't get ingested.

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py [hunk 14]
```diff
@@ -0,0 +1,283 @@
...
+
+            # temp check to prevent existing degree/programs from being overwritten
+            additioanl_metadata = DegreeAdditionalMetadata.objects.filter(degree=degree).first()
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/data_loaders/degrees_loader.py]: nit: additional_metadata

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py [hunk 15]
```diff
@@ -0,0 +1,283 @@
...
+                ))
+                self.messages_list.append('[DEGREE UPDATE ERROR] degree {}'.format(
+                    degree_title
+                ))
+                continue
+
+            logger.info("Degree {} {} located in the database. {} degree.".format(   # lint-amnesty, pylint: disable=logging-format-interpolation
+                degree_title,
+                "is" if degree else "is not",
+                "Creating new" if not degree else "Updating existing"
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/data_loaders/degrees_loader.py:87]: 🥇

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py [hunk 16]
```diff
@@ -0,0 +1,283 @@
...
+            ))
+
+            try:
+                degree = self._update_or_create_degree(row, program_type)
+            except Exception:   # pylint: disable=broad-except
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/data_loaders/degrees_loader.py:97]: Since this is a broad exception, we can add a comment here to explain what sort of exceptions can be expected here.

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py [hunk 17]
```diff
@@ -0,0 +1,283 @@
...
+                    "updating" if degree else "creating"
+                ))
+                self.messages_list.append('[DEGREE {} ERROR] degree {}'.format(
+                    "UPDATE" if degree else "CREATE",
+                    degree_title
+                ))
+                continue
+
+            # degree.authoring_organizations.clear()
+            degree.authoring_organizations.add(org)
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/data_loaders/degrees_loader.py]: We can move this to a method too, for the sake of consistency.

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py [hunk 18]
```diff
@@ -0,0 +1,283 @@
...
+
+            self._handle_additional_metadata(row, degree)
+            self._handle_image_fields(row, degree)
+            self._handle_specializations(row, degree)
+            self._handle_courses(row, degree)
+            # TODO: handle curricula
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/data_loaders/degrees_loader.py]: to remove this TODO comment.

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py [hunk 19]
```diff
@@ -0,0 +1,283 @@
...
+            transformed_dict[updated_key] = value
+        return transformed_dict
+
+    def _update_or_create_degree(self, data, program_type):
+        """
+        Make a degree object through ORM
+        """
+        data_dict = {
+            "type": program_type,
+            "status": ProgramStatus.Unpublished,
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/data_loaders/degrees_loader.py:163]: For now, it is ok. But for the future, we would want to use the status of the degree, falling back to Unpublished if a new degree is created.

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py [hunk 20]
```diff
@@ -0,0 +1,283 @@
...
+            # "primary_subject_override":  primary_subject_override,
+            # "level_type_override": level_type_override,
+            # "language_override": language_override,
+            "marketing_slug": data['slug'],
+            "overview": data['overview'],
+            # "organization_short_code_override": data.get('organization_short_code_override', ''),
+            "partner": self.partner,
+
+        }
+        # TODO: handle exceptions
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/data_loaders/degrees_loader.py]: to remove this todo, as the exception will be handled by the caller.

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py [hunk 21]
```diff
@@ -0,0 +1,283 @@
...
+        degree, updated = Degree.objects.update_or_create(
+            title=data['title'],
+            # additional_metadata__external_identifier=data['identifier'],
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/data_loaders/degrees_loader.py]: Reminder to activate this check.

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py [hunk 22]
```diff
@@ -0,0 +1,283 @@
...
+        if data.get('organization_logo_override'):
+            is_downloaded = download_and_save_program_image(
+                program, data['organization_logo_override'],
+                'organization_logo_override'
+            )
+            if not is_downloaded:
+                logger.error("Unexpected error happened while downloading image for degree {}".format(  # lint-amnesty, pylint: disable=logging-format-interpolation
+                    degree.title
+                ))
+                self.messages_list.append('[IMAGE DOWNLOAD FAILURE] degree {}'.format(degree.title))
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/data_loaders/degrees_loader.py]: OVERRIDE IMAGE DOWNLOAD FAILURE

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py [hunk 23]
```diff
@@ -0,0 +1,283 @@
...
+
+            program = Program.objects.get(degree=degree, partner=self.partner)
+            curriculam, created = Curriculum.objects.update_or_create(
+                program=program,
+                defaults={
+                    'marketing_text': marketing_text,
+                }
+            )
+
+            logger.info("Curriculam {} is {} with Degree title {}".format(    # lint-amnesty, pylint: disable=logging-format-interpolation
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/data_loaders/degrees_loader.py]: nit: for/against Degree title

File: course_discovery/apps/course_metadata/data_loaders/tests/mixins.py [hunk 24]
```diff
@@ -42,6 +42,116 @@ def test_init(self):
...
+        'organic_url',
+        'overview',
+        'specializations',
+        'courses',
+    ]
+
+    # TODO: update it
+    MINIMAL_CSV_DATA_KEYS_ORDER = CSV_DATA_KEYS_ORDER
+
+    # TODO: handle card image url
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/data_loaders/tests/mixins.py]: to remove the TODOs

File: course_discovery/apps/course_metadata/data_loaders/tests/test_degree_csv_loader.py [hunk 25]
```diff
@@ -0,0 +1,273 @@
...
+                degree = Degree.objects.get(title=self.DEGREE_TITLE, partner=self.partner)
+                program = Program.objects.get(degree=degree, partner=self.partner)
+                curriculam = Curriculum.objects.get(program=program)
+
+                assert degree.card_image.read() == image_content
+                assert degree.specializations.count() == 2
+                assert curriculam.marketing_text == "ABC\nD&E\nHarvard CS50"
+                self._assert_degree_data(degree, self.BASE_EXPECTED_DEGREE_DATA)
+
+    def test_ingest_flow_for_preexisting_course(self, jwt_decode_patch):  # pylint: disable=unused-argument
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/data_loaders/tests/test_degree_csv_loader.py]: nit: preexisting_degree

File: course_discovery/apps/course_metadata/data_loaders/tests/test_degree_csv_loader.py [hunk 26]
```diff
@@ -0,0 +1,273 @@
...
+                degree = Degree.objects.get(title=self.DEGREE_TITLE, partner=self.partner)
+                program = Program.objects.get(degree=degree, partner=self.partner)
+                curriculam = Curriculum.objects.get(program=program)
+
+                assert degree.specializations.count() == 2
+                assert curriculam.marketing_text == "ABC\nD&E\nHarvard CS50"
+                assert degree.card_image.read() == image_content
+                self._assert_degree_data(degree, self.BASE_EXPECTED_DEGREE_DATA)
+
+    def test_ingest_flow_for_minimal_course_data(self, jwt_decode_patch):  # pylint: disable=unused-argument
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/data_loaders/tests/test_degree_csv_loader.py]: minimal_degree_data

**Review by DawoudSheraz — COMMENTED** [review]
- to add unit tests for import_degree_data mgt command
- Nits comments
A very nice PR overall.

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py (hunk 11 above)
Comment (azanbinzahid) [reply course_discovery/apps/course_metadata/data_loaders/degrees_loader.py]: I don't think so. We will have clear logs on what went wrong.

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py [hunk 27]
```diff
@@ -0,0 +1,285 @@
...
+
+    def __init__(self, partner, api_url=None, max_workers=None, is_threadsafe=False, csv_path=None):
+        super().__init__(partner, api_url, max_workers, is_threadsafe)
+
+        self.messages_list = []  # to show failure/skipped ingestion message at the end
+        self.degree_uuids = {}  # to show the discovery degrees/program ids for each processed degree
+        try:
+            self.reader = csv.DictReader(open(csv_path, 'r'))  # lint-amnesty, pylint: disable=consider-using-with
+        except FileNotFoundError:
+            logger.exception("Error opening csv file at path {}".format(csv_path))    # lint-amnesty, pylint: disable=logging-format-interpolation
```
Comment (Ali-D-Akbar) [inline course_discovery/apps/course_metadata/data_loaders/degrees_loader.py:29]: Is there a reason why we're not using f-string in logs and other places in this file?

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py [hunk 28]
```diff
@@ -0,0 +1,299 @@
...
+                continue
+
+            # get degree object from title and external_identifier
+            degree = Degree.objects.filter(
+                title=degree_title, partner=self.partner,
+                # additional_metadata__external_identifier=row['identifier']
+            ).first()
+
+            # temp check to prevent existing degree/programs from being overwritten
+            additional_metadata = DegreeAdditionalMetadata.objects.filter(degree=degree).first()
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/data_loaders/degrees_loader.py:74]: Expecting this to be removed once there is a verdict on title uniqueness.

Comment (azanbinzahid) [reply course_discovery/apps/course_metadata/data_loaders/degrees_loader.py:74]: Yes, its a temp check.

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py [hunk 29]
```diff
@@ -0,0 +1,299 @@
...
+            updated_key = key.strip().lower().replace(' ', '_')
+            transformed_dict[updated_key] = value
+        return transformed_dict
+
+    def _update_or_create_degree(
+        self, data, program_type, primary_subject_override,
+        level_type_override, language_override
+    ):
+        """
+        Make a degree object through ORM
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/data_loaders/degrees_loader.py]:
```suggestion
        Create or Update a degree object through ORM
```

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py [hunk 30]
```diff
@@ -0,0 +1,299 @@
...
+
+        program = Program.objects.get(degree=degree, partner=self.partner)
+        is_downloaded = download_and_save_program_image(
+            program, data['card_image_url'],
+        )
+        if not is_downloaded:
+            logger.error("Unexpected error happened while downloading image for degree {}".format(  # lint-amnesty, pylint: disable=logging-format-interpolation
+                degree.title
+            ))
+            self.messages_list.append('[OVERRIDE IMAGE DOWNLOAD FAILURE] degree {}'.format(degree.title))
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/data_loaders/degrees_loader.py]: nit: Degree Image download failure

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py [hunk 31]
```diff
@@ -0,0 +1,299 @@
...
+
+            org = self._get_object(Organization, "key", row['organization_key'], degree_title)
+            program_type = self._get_object(ProgramType, "slug", program_type, degree_title)
+            primary_subject_override = self._get_object(
+                Subject, "translations__name",
+                row['primary_subject'], degree_title
+            )
+            level_type_override = self._get_object(
+                LevelType, "translations__name_t",
+                row['course_level'], degree_title
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/data_loaders/degrees_loader.py:49]: Ok for now, but the header here is incorrectly referencing Course Level. The header in CSV should be Degree Level

Comment (azanbinzahid) [reply course_discovery/apps/course_metadata/data_loaders/degrees_loader.py:49]: Will keep it as it is then.

**Review by DawoudSheraz — APPROVED** [review]

**Review by AliAdnanSohail — APPROVED** [review]
Looks good to me. Great work 👏

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py [hunk 32]
```diff
@@ -0,0 +1,299 @@
...
+    Curriculum, Degree, DegreeAdditionalMetadata, LanguageTag, LevelType, Organization, Program, ProgramType,
+    Specialization, Subject
+)
+from course_discovery.apps.course_metadata.utils import download_and_save_program_image
+
+logger = logging.getLogger(__name__)
+
+
+class DegreeCSVDataLoader(AbstractDataLoader):
+
```
Comment (Ali-D-Akbar) [inline course_discovery/apps/course_metadata/data_loaders/degrees_loader.py:20]: missing docstring for class.

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py [hunk 33]
```diff
@@ -0,0 +1,299 @@
...
+        if data.get('organization_logo_override'):
+            is_downloaded = download_and_save_program_image(
+                program, data['organization_logo_override'],
+                'organization_logo_override',
+            )
+            if not is_downloaded:
+                logger.error("Unexpected error happened while downloading image for degree {}".format(  # lint-amnesty, pylint: disable=logging-format-interpolation
+                    degree.title
+                ))
+                self.messages_list.append('[OVERRIDE IMAGE DOWNLOAD FAILURE] degree {}'.format(degree.title))
```
Comment (Ali-D-Akbar) [inline course_discovery/apps/course_metadata/data_loaders/degrees_loader.py]: nit: `ORG LOGO OVERRIDE IMAGE DOWNLOAD FAILURE`

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py [hunk 34]
```diff
@@ -0,0 +1,299 @@
...
+            ))
+            self.messages_list.append('[DEGREE IMAGE DOWNLOAD FAILURE] degree {}'.format(degree.title))
+
+        if data.get('organization_logo_override'):
+            is_downloaded = download_and_save_program_image(
+                program, data['organization_logo_override'],
+                'organization_logo_override',
+            )
+            if not is_downloaded:
+                logger.error("Unexpected error happened while downloading image for degree {}".format(  # lint-amnesty, pylint: disable=logging-format-interpolation
```
Comment (Ali-D-Akbar) [inline course_discovery/apps/course_metadata/data_loaders/degrees_loader.py]: nit: org logo image

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py [hunk 35]
```diff
@@ -0,0 +1,299 @@
...
+
+            program = Program.objects.get(degree=degree, partner=self.partner)
+            curriculam, created = Curriculum.objects.update_or_create(
+                program=program,
+                defaults={
+                    'marketing_text': marketing_text,
+                }
+            )
+
+            logger.info("Curriculam {} is {} for Degree title {}".format(    # lint-amnesty, pylint: disable=logging-format-interpolation
```
Comment (Ali-D-Akbar) [inline course_discovery/apps/course_metadata/data_loaders/degrees_loader.py]: nit: *Curriculum

File: course_discovery/apps/course_metadata/management/commands/tests/test_import_degree_data.py [hunk 36]
```diff
@@ -0,0 +1,91 @@
...
+from course_discovery.apps.course_metadata.data_loaders.tests import mock_data
+from course_discovery.apps.course_metadata.data_loaders.tests.mixins import DegreeCSVLoaderMixin
+from course_discovery.apps.course_metadata.models import Curriculum, Degree, Program
+
+LOGGER_PATH = 'course_discovery.apps.course_metadata.management.commands.import_degree_data'
+
+
+@mock.patch(
+    'course_discovery.apps.course_metadata.data_loaders.configured_jwt_decode_handler',
+    return_value={'preferred_username': 'test_username'}
```
Comment (Ali-D-Akbar) [inline course_discovery/apps/course_metadata/management/commands/tests/test_import_degree_data.py:22]: Just wondering, what's the purpose of this mock? and where does this `test_username` come from?

**Review by Ali-D-Akbar — APPROVED** [review]
Some nits and questions. Good job. 👍

**Comment by davidjoy** [comment]
This all looks reasonable to me.  I'm not going to be useful reviewing the details, but can ask some high-level questions:

1. Is this using a degrees data model that we already had in course-discovery? 
2. How close are we to what our external data source (the 2U degrees business) needs?  Do they have concepts or fields we don't have today?
3. Are there any opportunities here to improve our data model given that they, presumably, have a robust degree model of their own, and given the above?  I'm not pre-supposing either way; I just want to ask the question.  
4. Are we ensuring that, after questions 1-3 here are answered, that we're arriving at a data model that feels "standardized" or "normalized"?  

These are very vague questions, I know.  The point is that I'm curious to understand whether or not this is a good opportunity to make any improvements to our degrees data model before we start using it more heavily.

**Review by davidjoy — COMMENTED** [review]
This all looks reasonable to me.  I'm not going to be useful reviewing the details, but can ask some high-level questions:

1. Is this using a degrees data model that we already had in course-discovery? 
2. How close are we to what our external data source (the 2U degrees business) needs?  Do they have concepts or fields we don't have today?
3. Are there any opportunities here to improve our data model given that they, presumably, have a robust degree model of their own, and given the above?  I'm not pre-supposing either way; I just want to ask the question.  
4. Are we ensuring that, after questions 1-3 here are answered, that we're arriving at a data model that feels "standardized" or "normalized"?  

These are very vague questions, I know.  The point is that I'm curious to understand whether or not this is a good opportunity to make any improvements to our degrees data model before we start using it more heavily.

File: course_discovery/apps/course_metadata/data_loaders/degrees_loader.py (hunk 27 above)
Comment (azanbinzahid) [reply course_discovery/apps/course_metadata/data_loaders/degrees_loader.py:29]: Using f-strings in log statement causes pylint quality failure (https://pylint.pycqa.org/en/latest/messages/warning/logging-fstring-interpolation.html)

File: course_discovery/apps/course_metadata/management/commands/tests/test_import_degree_data.py (hunk 36 above)
Comment (azanbinzahid) [reply course_discovery/apps/course_metadata/management/commands/tests/test_import_degree_data.py:22]: Since we are using `AbstractDataLoader` as base class, it requires user to be logged in.

**Comment by azanbinzahid** [comment]
Thank you @davidjoy for raising these questions. Here are in-order answers:

1. Yes, we are using the same model.
2. We have mapped all of field that was in the external source (as of now). New fields/models have been created and can be extended as needed. Details can be found in this [sheet](https://docs.google.com/spreadsheets/d/1EnijXN-sWyMvc8mEySTC1DxBCG0-qZfi5MJMvLofacs/edit#gid=374182909) as well.
3. We can't say no to improvements but NoSQL could be an answer for future/in coming dynamic fields. Although it would be a huge shift.  
4. For current business lines, yes. Else answer to question 3 as we can't really predict.

**Comment by davidjoy** [comment]
Thanks @azanbinzahid!

Just to be clear, when you say "NoSQL" do you mean adding something like a JSON field to the model for dynamic fields?  NoSQL makes me think we're talking about using a different database. 😄

**Comment by azanbinzahid** [comment]
Ah, we thought of JSON field first but disregarded it. I am suggesting a different database (obviously for longer-run) that would reduce the number of JOINS and queries and effort to follow a strict-structured approach.

**Comment by DawoudSheraz** [comment]
I would advocate the JSON field for dynamic fields like it is already present for Courseware Student Module in the platform. If we move to actual NoSQL, we would be in for a lot of tricky operations up our sleeve, not to mention the performance aspects.
